File: system-management/backend/src/main/java/com/darkness/system/management/service/GroupService.java
package com.darkness.system.management.service;

import com.darkness.system.management.domain.Group;
import com.darkness.system.management.domain.GroupMember;
import com.darkness.system.management.dto.request.CreateGroupRequest;
import com.darkness.system.management.dto.request.UpdateGroupRequest;
import com.darkness.system.management.dto.response.GroupResponse;
import com.darkness.system.management.dto.response.PageResponse;
import com.darkness.system.management.exception.DuplicateNameException;
import com.darkness.system.management.exception.ResourceNotFoundException;
import com.darkness.system.management.mapper.GroupMapper;
import com.darkness.system.management.repository.GroupMemberRepository;
import com.darkness.system.management.repository.GroupRepository;
import com.darkness.system.management.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
@RequiredArgsConstructor
public class GroupService {

    private final GroupRepository groupRepository;
    private final GroupMemberRepository groupMemberRepository;
    private final UserRepository userRepository;
    private final GroupMapper groupMapper;

    @Transactional(readOnly = true)
    public PageResponse<GroupResponse> listGroups(String search, Pageable pageable) {
        var page = (search == null || search.isBlank())
                ? groupRepository.findAll(pageable).map(groupMapper::toResponse)
                : groupRepository.findByNameContainingIgnoreCase(search, pageable).map(groupMapper::toResponse);
        return PageResponse.from(page);
    }

    @Transactional(readOnly = true)
    public GroupResponse getGroup(UUID groupId) {
        return groupMapper.toResponse(findOrThrow(groupId));
    }

    @Transactional
    public GroupResponse createGroup(CreateGroupRequest request) {
        if (groupRepository.existsByNameIgnoreCase(request.name())) {
            throw new DuplicateNameException("Group name already exists: " + request.name());
        }
        Group group = new Group();
        group.setName(request.name());
        group.setDescription(request.description());
        return groupMapper.toResponse(groupRepository.save(group));
    }

    @Transactional
    public GroupResponse updateGroup(UUID groupId, UpdateGroupRequest request) {
        Group group = findOrThrow(groupId);
        if (request.name() != null) {
            if (!request.name().equalsIgnoreCase(group.getName())
                    && groupRepository.existsByNameIgnoreCase(request.name())) {
                throw new DuplicateNameException("Group name already exists: " + request.name());
            }
            group.setName(request.name());
        }
        if (request.description() != null) group.setDescription(request.description());
        return groupMapper.toResponse(groupRepository.save(group));
    }

    @Transactional
    public void deleteGroup(UUID groupId) {
        if (!groupRepository.existsById(groupId)) {
            throw new ResourceNotFoundException("Group not found: " + groupId);
        }
        groupRepository.deleteById(groupId);
    }

    @Transactional
    public void addMember(UUID groupId, UUID userId) {
        findOrThrow(groupId);
        if (!userRepository.existsById(userId)) {
            throw new ResourceNotFoundException("User not found: " + userId);
        }
        if (groupMemberRepository.existsByIdUserIdAndIdGroupId(userId, groupId)) {
            return;
        }
        GroupMember member = new GroupMember();
        GroupMember.GroupMemberId id = new GroupMember.GroupMemberId();
        id.setUserId(userId);
        id.setGroupId(groupId);
        member.setId(id);
        groupMemberRepository.save(member);
    }

    @Transactional
    public void removeMember(UUID groupId, UUID userId) {
        findOrThrow(groupId);
        groupMemberRepository.deleteByIdUserIdAndIdGroupId(userId, groupId);
    }

    private Group findOrThrow(UUID groupId) {
        return groupRepository.findById(groupId)
                .orElseThrow(() -> new ResourceNotFoundException("Group not found: " + groupId));
    }
}

File: system-management/frontend/src/app/core/services/group.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { CreateGroupRequest, GroupResponse, UpdateGroupRequest } from '../models/group.model';
import { PageResponse } from '../models/user.model';

@Injectable({ providedIn: 'root' })
export class GroupService {
  private readonly apiUrl = `${environment.apiUrl}/groups`;

  constructor(private http: HttpClient) {}

  listGroups(page = 0, size = 10, search?: string): Observable<PageResponse<GroupResponse>> {
    let params = new HttpParams().set('page', page).set('size', size);
    if (search) params = params.set('search', search);
    return this.http.get<PageResponse<GroupResponse>>(this.apiUrl, { params });
  }

  getGroup(id: string): Observable<GroupResponse> {
    return this.http.get<GroupResponse>(`${this.apiUrl}/${id}`);
  }

  createGroup(request: CreateGroupRequest): Observable<GroupResponse> {
    return this.http.post<GroupResponse>(this.apiUrl, request);
  }

  updateGroup(id: string, request: UpdateGroupRequest): Observable<GroupResponse> {
    return this.http.patch<GroupResponse>(`${this.apiUrl}/${id}`, request);
  }

  deleteGroup(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  addMember(groupId: string, userId: string): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${groupId}/members/${userId}`, {});
  }

  removeMember(groupId: string, userId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${groupId}/members/${userId}`);
  }
}
